### PR TITLE
fix(clickhouse): bound system sampling reads instead of failing on max_rows_to_read

### PR DESCRIPTION
**Pull request by mikebridge**
### SUMMARY

ClickHouse enforces `max_rows_to_read` against its **pre-execution row estimate** derived from index/part analysis, which ignores `LIMIT`. Any full-table-shaped query on a table above the configured cap (observed: a ~4.3B-row table vs a 4B cap on ClickHouse Cloud) is rejected up front with `Code: 158 (TOO_MANY_ROWS)` — no matter how small the `LIMIT` is.

Superset authors several **system-generated** sampling/metadata queries with exactly this shape, all of which fail on large ClickHouse tables:

- **Filter-values dropdown** (`values_for_column`): `SELECT DISTINCT col FROM table LIMIT n` — user-facing error when it fails.
- **Samples tab / dataset preview** (samples query action): `SELECT * FROM table LIMIT n` — user-facing error.
- **Datetime format detection** on dataset save: `SELECT col FROM table LIMIT n` — failure is caught, but detection silently never works and the logs report success anyway.

This PR adds an engine-spec hook pair — `apply_sampling_read_limit_override` and `is_read_limit_error` — used to **retry** a failed system sampling query with a bounded read. The query always runs unchanged first; only when the engine rejects it with a read-limit error is it retried once with `SETTINGS read_overflow_mode='break'` appended, which makes the operator's own `max_rows_to_read` cap the read *bound*: the retry stops reading at the cap and returns the partial result instead of erroring. Both hooks are no-ops for other engines.

Design decisions:

- **Retry-only, never up-front.** Deployments whose sampling queries already succeed never see altered SQL. This matters for ClickHouse connections provisioned with `readonly=1`, which reject in-query SETTINGS changes (`Code: 164 READONLY`) — verified against a live ClickHouse 25.8 with a `readonly=1` user. If the retry itself fails (e.g. readonly), the **original** read-limit error is surfaced, so such deployments behave exactly as before this change.
- **Only Superset-authored statements over physical tables are retried.** Virtual datasets and any user-authored SQL stay fully governed by the operator's read limits. The samples query action marks its query object with a `system_sampling` extras flag consumed at execution time; the zero-row adhoc type probe is explicitly excluded (covered separately by #42298). `Database.select_star` output is never modified — it is displayed in SQL Lab, returned by the API, and persisted as CTAS result-fetch SQL, so it must stay clean.
- **Per-database opt-out**: operators can set `disable_sampling_read_limit_override: true` in the database `extra` to disable the retry entirely and keep read-limit rejections as hard failures. All call sites go through `Database.run_with_sampling_read_limit_retry` / `sampling_read_limit_retry_sql` so the opt-out is honored in one place.
- **The retry operates on the final mutated statement** (after `SQL_QUERY_MUTATOR`), so the `SETTINGS` clause is appended on its own line — safe after mutator-appended trailing comments — and a trailing statement terminator is stripped. Verified valid against ClickHouse 25.8.
- Read-limit classification (`is_read_limit_error`) is anchored to ClickHouse's error-code tokens (`Code: 158` / `TOO_MANY_ROWS`), not the setting name, so errors merely mentioning `max_rows_to_read` are not misclassified.

Also makes the datetime format detector truthful about failures:

- A read-limit rejection (after the retry is exhausted) logs at WARNING instead of ERROR — it is an expected, handled condition (detection is skipped for that column).
- The summary line reported `Detected formats for %d columns` using the count of *attempted* columns, so a 0-for-4 failure logged as apparent full success. It now reports `Detected formats for %d of %d temporal columns` and logs at WARNING when nothing was detected.

An UPDATING.md entry documents the behavior change (read-limit rejections on system sampling queries become bounded partial results by default) and the opt-out. Note the intentional trade-off: a retried filter-values query returns a partial value set with no truncation indicator — previously it errored loudly; the opt-out restores hard failure for operators who prefer it.

#### Follow-ups (out of scope here)

- **UI toggle for the opt-out**: expose `disable_sampling_read_limit_override` as a checkbox in the database modal's Extra Options (its siblings `disable_data_preview` / `disable_drill_to_detail` have one); until then it is set via the Extra JSON field.
- **Dedicated `QueryObject` attribute for the `system_sampling` marker**: the marker currently rides in the `extras` dict and is non-injectable via `/api/v1/chart/data` only because `ChartDataExtrasSchema` rejects unknown keys by default; a first-class attribute (or a schema test pinning that invariant) would make the boundary explicit.
- **Partial-result indicator**: a bounded-read retry returns partial data (e.g. an incomplete filter-values list) with no user-facing truncation signal.
- **Other engines**: implement the hook pair for engines with comparable pre-execution read limits as they are observed.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Captured against a live ClickHouse 25.8 (1M-row MergeTree table, connecting user capped at `max_rows_to_read=100000`), opening a filter on the `get_end` column in Explore — which fires the `values_for_column` `SELECT DISTINCT` query.

**Before** — the values query is rejected up front (`Code: 158 ... max rows: 100.00 thousand, current rows: 1.00 million (TOO_MANY_ROWS)`, endpoint returns 500) and the dropdown stays empty:

before: filter values dropdown empty, values endpoint 500 TOO_MANY_ROWS

**After** — the first attempt is rejected identically, the bounded-read retry kicks in (`WARNING ... retrying with the engine's bounded-read override`), and the dropdown populates with the 10,000 values from the capped read (endpoint returns 200):

after: filter values dropdown populated with 10000 options via bounded read

### TESTING INSTRUCTIONS

Unit tests cover the engine-spec hooks (base no-ops, ClickHouse family, idempotency, terminator stripping, trailing-comment safety, error classification), the retry helper (success never alters SQL, retry on read-limit, other errors re-raised, original error surfaced when the retry fails, opt-out honored), the call sites (filter values, samples action, format detector), the physical-vs-virtual distinction, `select_star` staying unmodified, the type-probe exclusion, and the detector logging changes:

```
pytest tests/unit_tests/db_engine_specs/test_clickhouse.py \
       tests/unit_tests/models/helpers_test.py \
       tests/unit_tests/datasets/test_datetime_format_detector.py \
       tests/unit_tests/common/test_query_actions.py
```

The retry semantics were additionally verified end-to-end against a live ClickHouse 25.8 (1M-row MergeTree table, user capped at `max_rows_to_read=100000`, plus a `readonly=1` user): direct over-cap sampling fails with code 158; the retry returns bounded results; the newline `SETTINGS` clause is valid including after trailing comments; the readonly user's under-cap queries run with unaltered SQL; and a readonly over-cap query surfaces the original TOO_MANY_ROWS error.

Manual verification:

1. Connect a ClickHouse database and pick a table whose row count exceeds `max_rows_to_read` (or set `max_rows_to_read` below the table's row count to simulate).
2. Create a physical dataset on it: dataset save completes and datetime format detection logs detected-of-attempted counts instead of `TOO_MANY_ROWS` errors.
3. Open a filter on a column of that dataset: the values dropdown populates (bounded read) instead of erroring.
4. Open the dataset's Samples tab: rows are returned.
5. Set `"disable_sampling_read_limit_override": true` in the database's `extra` and confirm the queries fail with `TOO_MANY_ROWS` again.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by mikebridge** [comment]
## Review

**What it does:** ClickHouse rejects Superset's system-generated sampling queries (filter-values dropdown, Samples tab, datetime format detection) up front when a table's *estimated* row count exceeds `max_rows_to_read`, even though the query has a `LIMIT`. This PR adds an engine-spec hook pair (`apply_sampling_read_limit_override` / `is_read_limit_error`) so those specific queries retry once with `SETTINGS read_overflow_mode='break'` on rejection, returning a bounded partial result instead of erroring. Retry-only (never up-front), physical-tables-only, and per-database opt-outable via `disable_sampling_read_limit_override`.

### Strengths
- **Well-scoped blast radius.** Retry only fires after the unmodified query fails, so deployments where sampling already works (including ClickHouse `readonly=1` connections, which reject in-query `SETTINGS`) see zero behavior change.
- **Consistent with existing prior art.** The hook-pair pattern mirrors `get_column_description_retry_sql` already in `base.py`.
- **Correctly handles the shallow-copy gotcha** in `_get_samples` — rebuilds the `extras` dict instead of mutating it in place, and it's directly tested (`test_get_samples_marks_query_as_system_sampling`).
- **Test coverage is thorough**: idempotency, terminator stripping, trailing-comment safety, error-code anchoring, retry-fails-surfaces-original-error, opt-out, virtual-vs-physical routing, and `select_star`/type-probe exclusions are all covered.
- **Honest logging fix** — the `detect_all_formats` summary previously reported "success" on a 0-for-4 run; now it's accurate and demoted to WARNING when nothing was detected.

### Issues worth a look

1. **Minor API inconsistency vs. the existing sibling hook** — `get_column_description_retry_sql` returns `str | None` (`None` = no retry). `apply_sampling_read_limit_override` instead returns the SQL *unchanged* as its "no override" sentinel, forcing callers (`sampling_read_limit_retry_sql`) to detect that via string equality (`retry_sql != sql`). Returning `Optional[str]` would match the existing convention and avoid any theoretical false-negative from string-equal-but-differently-formatted SQL.

2. **No handling for a pre-existing `SETTINGS` clause.** `ClickHouseBaseEngineSpec.apply_sampling_read_limit_override` blindly appends `\nSETTINGS read_overflow_mode='break'`. If the final mutated SQL already contains a `SETTINGS ...` clause from some other source, ClickHouse only permits one per statement — the retry would produce invalid SQL. Narrow edge case given it's system-generated SQL, but worth a defensive check or comment.

3. **The `system_sampling` extras marker is a soft boundary.** It's only safe from user injection today because `ChartDataExtrasSchema` happens to reject unknown keys by default, not an explicit check. The description already lists a first-class `QueryObject` attribute (or a pinning schema test) as a follow-up — given this marker controls whether a query bypasses the operator's read-limit-as-hard-failure semantics, I'd consider pulling the schema-pinning test into this PR rather than a someday-follow-up, since an unrelated future change to `ChartDataExtrasSchema` could silently reopen it.

4. **Minor pre-existing nit, not introduced here:** the docstring list in `superset/databases/schemas.py` concatenates `"...datasets."` + `"9. The \`\`disable_sampling_read_limit_override\`\`..."` with no separating space/newline — same issue already present between items 6→7 and 7→8, so not a regression.

### Overall
Solid, carefully-scoped fix with strong test coverage and honest documentation of trade-offs. Nothing here blocks merging — items 1 and 4 are cosmetic, item 2 is a narrow edge case, and item 3 is already on the radar as a follow-up but worth considering pulling forward given it's a security-relevant boundary.

---
*Posted via automated review request in #eng-reviews.*

**Comment by mikebridge** [comment]
Thanks for the review — addressed in f73f21fac8:

1. **API consistency**: `apply_sampling_read_limit_override` now returns `str | None`, matching the `get_column_description_retry_sql` convention. `None` means no retry; the string-equality sentinel in `Database.sampling_read_limit_retry_sql` is gone.
2. **Pre-existing `SETTINGS` clause**: the ClickHouse implementation now returns `None` when the SQL already contains a `SETTINGS` clause from any source (this also subsumes the previous own-suffix idempotency check). A subquery `SETTINGS` would match too, but the failure mode there is just "no retry" — i.e. the engine's normal rejection — never invalid SQL. Covered by `test_sampling_read_limit_override_existing_settings_returns_none`.
3. **Schema pinning pulled into this PR**: `test_chart_data_extras_rejects_system_sampling` now asserts `ChartDataExtrasSchema` rejects the `system_sampling` key, so a future `unknown = INCLUDE` (or explicit field) can't silently make the marker client-settable. The first-class `QueryObject` attribute remains a follow-up.
4. The `schemas.py` docstring spacing is left as-is since the same pattern spans items 6→9 and fixing only item 9 would be inconsistent — happy to do a separate cleanup of the whole list if wanted.

_This comment was generated by Claude (AI) on behalf of @mikebridge._

**Review by aminghadersohi — COMMENTED** [review]
Reviewed at `f73f21f`. Carefully-scoped fix with strong, non-vacuous test coverage and honest documentation of the trade-offs. Cruxes check out and the latest commit already resolved the return-sentinel, double-SETTINGS, and schema-pin concerns from your own posted review. Findings below are all LOW/nit — nothing blocking. (Posting as a comment for a maintainer to approve.)

### Crux verdicts
- **Resource bound safe & applied to all paths** — Superset computes no numeric bound (so no falsy-zero pitfall); the bound is ClickHouse's own `max_rows_to_read` via `read_overflow_mode='break'`, which reads *fewer* rows and returns a partial result — it can't exceed the cap. Applied to all three sampling paths (`values_for_column`, `query()` samples, detector), retry-only, with graceful degradation (original error surfaced if the retry fails).
- **SQL construction injection-safe** — constant suffix, no user/column/table interpolation; identifier quoting unchanged.
- **base.py hook safe for other engines** — base returns `None`/`False`, so non-ClickHouse specs get no retry and zero behavior change.
- **datetime_format_detector** — no regression for non-ClickHouse: detection logic is unchanged; only the execution wrapper and logging changed, and both fall through to the prior behavior when the engine has no override.
- **`system_sampling` extras marker** — confirmed `ChartDataExtrasSchema` has no explicit `unknown` handling, so marshmallow's default `RAISE` blocks injection via `/api/v1/chart/data`; the new `test_chart_data_extras_rejects_system_sampling` pins that invariant. Note that even if it were injected, `read_overflow_mode='break'` reads *fewer* rows, so it is not a data-access escalation — good.

### Findings (all LOW / non-blocking)

1. **`\bSETTINGS\b` guard false-positives on a column named `settings`.** The new `re.search(r"\bSETTINGS\b", sql, re.IGNORECASE)` in `apply_sampling_read_limit_override` matches a bare identifier, so e.g. a filter-values query `SELECT DISTINCT settings AS column_values FROM t LIMIT n` (or datetime detection on a `settings` column) returns `None` and silently skips the retry — the fix never applies for that column. It fails safe (degrades to the prior `TOO_MANY_ROWS` error), but `settings` is a plausible column name. Consider anchoring the guard to a trailing/top-level SETTINGS clause rather than any occurrence of the token.

2. **`get_df` call sites append SETTINGS pre-mutation (informational).** For the samples and detector paths (which go through `Database.get_df`), the `\nSETTINGS …` suffix is appended to the SQL *before* `get_df` re-applies `SQL_QUERY_MUTATOR`; the "own-line survives a trailing comment" reasoning strictly holds only for `values_for_column` (pre-mutated + direct `read_sql_query`). Impact is limited — `SQL_QUERY_MUTATOR` is operator-controlled and a retry failure surfaces the original error — but a wrapping/subquery-style mutator would misplace the clause. Worth a one-line note.

3. **Nit:** in `superset/databases/schemas.py` the docstring concatenates `"…datasets."` and `"9. The ``disable_sampling_read_limit_override``…"` with no separating space (same as the existing 6→7 and 7→8 joins).

### Tests
All new test files are non-vacuous — each fails when its corresponding production change is reverted. The detector fixture re-implements the retry helper, but the real `Database.run_with_sampling_read_limit_retry` is exercised in `test_clickhouse.py`, and the detector tests still fail on reverting the detector wiring/logging, so the coverage is sound. Bound, opt-out, retry-fails-surfaces-original, physical-vs-virtual routing, and the `select_star`/type-probe exclusions are all covered.

### Scope
All six production files are in scope for the stated fix (three sampling call sites + the hook pair + the retry helper + the schema field); this reads as a cohesive change rather than scope-creep, and the `datetime_format_detector` change is justified as one of the three named sampling queries.

**Comment by mikebridge** [comment]
Thanks @aminghadersohi — all three findings addressed:

1. **`settings`-column false positive**: the guard now matches the clause shape `SETTINGS <key> = ...` rather than the bare token, so a column named `settings` (selected, aliased, or in a `WHERE settings = ...` predicate) no longer suppresses the retry. Pinned by `test_sampling_read_limit_override_ignores_settings_named_column`.
2. **Pre-mutation SETTINGS on `get_df` paths**: documented on `Database.run_with_sampling_read_limit_retry` — `run` implementations going through `get_df` re-apply `SQL_QUERY_MUTATOR` to the retry SQL; comment-adding (any position) and subquery-wrapping mutators are fine for ClickHouse, and an appending mutator degrades to the retry failing with the original read-limit error surfaced.
3. **Docstring joins**: restored the missing `<br/>` separators on items 6→9 (items 1–5 already had them), including the missing space inside item 7.

Also merged latest master, which grew an overlap: #42388 wraps the detector's DB query in a WARNING-level catch. The bounded-read retry now runs inside that catch, and this branch's narrower read-limit-only WARNING downgrade is superseded by upstream's broader all-DB-failures-warn behavior — the detector diff is smaller than before as a result.

_This comment was generated by Claude (AI) on behalf of @mikebridge._

**Comment by mikebridge** [comment]
cc @eschutho — you touched `superset/datasets/datetime_format_detector.py` in #42388 (WARNING-level logging fix) yesterday, which overlaps with the logging changes here. Would appreciate a look.

cc @joe-clickhouse — for a ClickHouse-domain review of the `max_rows_to_read` retry behavior.

(No formal code owner exists for these files per CODEOWNERS — I don't have permission to formally request reviewers on this repo, so flagging via mention instead.)

**Comment by mikebridge** [comment]
@eschutho heads-up: this branch overlapped with your #42388 in `datetime_format_detector.py` — both changed the `database.get_df(...)` call site. How the rebase resolved it, in case you want to sanity-check:

- Your `try/except` (all DB-query failures → WARNING + `return None`) is kept as the outer structure; this PR's bounded-read retry now runs **inside** it — `run_with_sampling_read_limit_retry(...)` wraps the `get_df`. So on ClickHouse a `max_rows_to_read` rejection first gets one retry with `SETTINGS read_overflow_mode='break'`; if that also fails, it falls into your catch and logs the WARNING like any other DB failure.
- This branch originally carried its own narrower read-limit-only WARNING downgrade in the outer `except` — that's now redundant with your broader catch, so it was dropped and the outer handler is back to plain `logger.exception`, reserved for genuine detection-logic bugs (matching your intent in #42388).
- Tests: your updated `test_detect_column_format_error_handling` and new `test_detect_column_format_internal_error_still_logs_at_error` are kept as-is; this PR's read-limit logging test now asserts your "Could not query column" WARNING message, and its old assert-ERROR-on-connection-failure test was removed since #42388 intentionally changed that behavior.

Net effect: your logging semantics are preserved exactly; this PR only adds the retry attempt before a failure reaches your catch.

_This comment was generated by Claude (AI) on behalf of @mikebridge._

File: tests/unit_tests/datasets/test_datetime_format_detector.py [hunk 1]
```diff
@@ -49,6 +50,23 @@ def mock_dataset() -> MagicMock:
...
+        message = str(ex)
+        return "TOO_MANY_ROWS" in message or "Code: 158" in message
+
+    def run_with_retry(sql: str, run: Any) -> Any:
+        try:
+            return run(sql)
+        except Exception as ex:
+            if not is_read_limit_error(ex):
+                raise
+            return run(f"{sql}\nSETTINGS read_overflow_mode='break'")
```
[the remark this answers is not in the document]
Comment (mikebridge) [reply tests/unit_tests/datasets/test_datetime_format_detector.py]: Valid — fixed in 8455828d67. The fixture's retry helper now raises the original read-limit error when the retry fails (`raise ex from retry_ex`), mirroring `Database.run_with_sampling_read_limit_retry`, and a new regression test (`test_detect_column_format_failed_retry_surfaces_original_error`) covers the exact readonly-retry-failure scenario: TOO_MANY_ROWS on the first attempt, Code: 164 READONLY on the retry, asserting the WARNING carries the original error and not the retry failure.

_This comment was generated by Claude (AI) on behalf of @mikebridge._

File: superset/db_engine_specs/clickhouse.py [hunk 2]
```diff
@@ -56,6 +56,45 @@ class ClickHouseBaseEngineSpec(BaseEngineSpec):
...
+        carries the override, or it contains a SETTINGS clause from another
+        source (ClickHouse permits only one per statement, so appending a
+        second would produce invalid SQL — including subquery SETTINGS in
+        this check merely degrades to the engine's normal rejection). The
+        guard matches the clause shape ``SETTINGS <key> = ...`` rather than
+        the bare token, so a column that happens to be named ``settings``
+        does not suppress the retry. A trailing statement terminator is
+        stripped so the SETTINGS clause attaches to the statement itself.
+        """
+        if re.search(r"\bSETTINGS\s+\w+\s*=", sql, re.IGNORECASE):
```
[the remark this answers is not in the document]
Comment (mikebridge) [reply superset/db_engine_specs/clickhouse.py]: Partially valid and fixed in 8455828d67 — worth noting the failure mode was fail-safe (a false positive suppresses the retry, degrading to the pre-fix TOO_MANY_ROWS error, never invalid SQL). The guard now blanks single-quoted string literals, double-quoted/backtick identifiers, and line/block comments before matching the clause shape, so clause-shaped text in a `fetch_values_predicate` literal or a mutator comment no longer suppresses the retry. Covered by `test_sampling_read_limit_override_ignores_settings_text_in_literals` (literal, comment, and genuine-clause cases). Full SQL parsing felt disproportionate for a guard whose miss is fail-safe.

_This comment was generated by Claude (AI) on behalf of @mikebridge._

File: superset/models/helpers.py [hunk 3]
```diff
@@ -3322,7 +3335,22 @@ def values_for_column(  # pylint: disable=too-many-locals
...
+                if not self.sql:
+                    # Physical-table dataset: the filter-values statement is
+                    # generated by Superset, so a read-limit rejection (e.g.
+                    # ClickHouse max_rows_to_read) is retried once with the
+                    # engine's bounded-read override. Virtual datasets embed
+                    # user-authored SQL and stay governed by operator read
+                    # limits.
+                    df = self.database.run_with_sampling_read_limit_retry(
+                        sql, run_query
+                    )
```
[the remark this answers is not in the document]
Comment (mikebridge) [reply superset/models/helpers.py:3360]: This one is the PR's documented, intended trade-off rather than an oversight — declining a behavior change here. Prior to this PR the dropdown *errored* on these tables, so users got zero values; a bounded partial list is strictly more useful, and the maintainer review on this PR reached the same verdict. The trade-off is called out in the PR description, in UPDATING.md (including the caching implication for operators), and `disable_sampling_read_limit_override` restores hard-failure semantics per database for operators who prefer them. A user-facing truncation indicator is tracked as an explicit follow-up (it needs a payload/UI contract change that would balloon this PR's scope). On the cache point specifically: a cached partial list behaves the same as any cached value list whose underlying data changed — it refreshes on the configured timeout.

_This comment was generated by Claude (AI) on behalf of @mikebridge._

**Review by aminghadersohi — COMMENTED** [review]
Re-reviewed at `15bf2ad0` (following the 5 commits since our last pass at `f73f21f`). All three findings from the previous review are resolved — verified in the code at this HEAD, not just from the commit messages:

1. **`settings`-named-column false positive** — the existing-clause guard is now anchored to the clause shape (`SETTINGS <key> = …`) and additionally blanks string literals, quoted identifiers, and comments before matching, so a column named `settings` (or clause-shaped text inside a literal/comment) no longer suppresses the bounded-read retry. Idempotency is preserved and it's covered by `test_sampling_read_limit_override_ignores_settings_named_column` / `..._ignores_settings_text_in_literals`. ✅
2. **`get_df` retry re-applies `SQL_QUERY_MUTATOR`** — now documented in the `run_with_sampling_read_limit_retry` docstring (which mutator shapes are supported, and that a failed retry surfaces the original read-limit error). ✅
3. **`schemas.py` docstring separators** — `<br/>` separators restored on items 6–8. ✅

The detector reconciliation with the newly-merged #42388 is clean: this PR simply slots `run_with_sampling_read_limit_retry` into #42388's existing try/except-at-WARNING structure and updates the summary to "detected of attempted" — it does not itself broaden error handling.

**Cruxes re-verified at this HEAD:** the bounded read cannot exceed the operator's cap (`read_overflow_mode='break'` stops at `max_rows_to_read` and returns partial — it reads *fewer* rows); the appended `SETTINGS` clause is a constant with no attacker-controlled interpolation and is now guarded against a duplicate clause; no raw exception text is surfaced to clients (WARNING logged server-side only); no function-body imports in production code; and the `system_sampling` extras marker remains non-injectable via `/api/v1/chart/data` (schema default-`RAISE`, pinned by `test_chart_data_extras_rejects_system_sampling`). Scope is unchanged — all touched files still map to the stated fix.

No new findings. The two outdated CodeAnt threads (detector test double; SETTINGS-over-raw-SQL) are genuinely addressed in the code at this HEAD. The one remaining live thread — a bounded filter-values retry returns a partial list cached without a truncation indicator — is the PR's documented, per-database opt-out-able (`disable_sampling_read_limit_override`) trade-off; I concur it's an intended behavior (strictly better than the prior hard error) rather than a defect, with the user-facing truncation signal appropriately tracked as a follow-up.

Leaving this as a comment for a maintainer to take the final approval decision.

**Review by aminghadersohi — APPROVED** [review]

**Review by eschutho — APPROVED** [review]
